sway-core/src/ir_generation/function.rs

```diff
@@ -2430,114 +2462,143 @@ impl<'eng> FnCompiler<'eng> {
-                let cur_type_info = &*cur_type_info_arc;
-                match (idx_kind, cur_type_info) {
-                    (
-                        ProjectionKind::StructField { name: idx_name },
-                        TypeInfo::Struct(decl_ref),
-                    ) => {
-                        let struct_decl = self.engines.de().get_struct(decl_ref);
-
-                        match struct_decl.get_field_index_and_type(idx_name) {
-                            None => {
-                                return Err(CompileError::InternalOwned(
-                                    format!(
-                                        "Unknown field name '{idx_name}' for struct '{}' \
-                                         in reassignment.",
-                                        struct_decl.call_path.suffix.as_str(),
-                                    ),
-                                    ast_reassignment.lhs_base_name.span(),
-                                ))
+        let lhs_ptr = match &ast_reassignment.lhs {
+            ty::TyReassignmentTarget::ElementAccess {
+                base_name,
+                base_type,
+                indices,
+            } => {
+                let name = self
+                    .lexical_map
+                    .get(base_name.as_str())
+                    .expect("All local symbols must be in the lexical symbol map.");
+
+                // First look for a local variable with the required name
+                let lhs_val = self
+                    .function
+                    .get_local_var(context, name)
+                    .map(|var| {
+                        self.current_block
+                            .append(context)
+                            .get_local(var)
+                            .add_metadatum(context, span_md_idx)
+                    })
+                    .or_else(||
+                        // Now look for an argument with the required name
+                        self.function
+                            .args_iter(context)
+                            .find_map(|(arg_name, arg_val)| (arg_name == name).then_some(*arg_val)))
+                    .ok_or_else(|| {
+                        CompileError::InternalOwned(
+                            format!("Variable not found: {name}."),
+                            base_name.span(),
+                        )
+                    })?;
+
+                if indices.is_empty() {
+                    // A non-aggregate; use a direct `store`.
+                    lhs_val
+                } else {
+                    // Create a GEP by following the chain of LHS indices. We use a scan which is
+                    // essentially a map with context, which is the parent type id for the current field.
+                    let mut gep_indices = Vec::<Value>::new();
+                    let mut cur_type_id = *base_type;
+                    // TODO-IG: Add support for projections being references themselves.
```

Comment:
This is an ongoing development of references tracked in #5063. The TODO is more as a bookmark to me to know where to look when starting the next step in the development. Otherwise, for bugs and issues unnrelated to ongoing development of references, issues are created and linked to TODOs in code.